lib/AtomicCache.php

```diff
@@ -0,0 +1,174 @@
+     * is invoked with the key as the first parameter and null as the second parameter. The value returned from the
+     * callback is stored in the cache and the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key) $create
+     * @param int|null $ttl
+     *
+     * @return Promise<mixed>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     */
+    public function load(string $key, callable $create, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $value = yield $this->cache->get($key);
+
+            if ($value !== null) {
+                return $value;
+            }
+
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                // Attempt to get the value again, since it may have been set while obtaining the lock.
+                $value = yield $this->cache->get($key);
+
+                if ($value !== null) {
+                    return $value;
+                }
+
+                return yield from $this->create($create, $key, null, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * The key is locked, the current value in the cache is accessed, then the $create callback is invoked with the key
+     * as the first parameter and the current value as the second parameter. The value returned from the callback is
+     * stored in cache and the promise returned from this method is resolved with the value.
+     *
+     * @param string   $key
+     * @param callable(string $key, mixed $value) $create
+     * @param int|null $ttl
+     *
+     * @return Promise<mixed>
+     *
+     * @throws CacheException If the $create callback throws an exception while generating the value.
+     */
+    public function swap(string $key, callable $create, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            $value = yield $this->cache->get($key);
+
+            try {
+                return yield from $this->create($create, $key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    private function lock(string $key): \Generator
+    {
+        try {
+            return yield $this->mutex->acquire($key);
+        } catch (\Throwable $exception) {
+            throw new CacheException(
+                \sprintf('Exception thrown when obtaining the lock for key "%s"', $key),
+                0,
+                $exception
+            );
+        }
+    }
+
+    private function create(callable $create, string $key, ?string $value, ?int $ttl = null): \Generator
```

Comment:
Are you sure about the `?string $value`? Above you call:

```php
$value = yield $this->cache->get($key);
//...
return yield from $this->create($create, $key, $value, $ttl);
```

But `Cache::get()` doesn't say that the cache value is a string.

Reply:
`Cache::set()` requires a string.

Reply:
Ah, I see. I'll send a PR later for `Cache::get()` to have `@return Promise<string|null>`.

Reply:
Can probably add resolution types on all the other methods too.

Reply:
Yeah, ofc. Which reminds me that `Promise` still lacks `@template` annotation: https://github.com/amphp/amp/pull/301